new lints around`#[must_use]`

llogiq opened the pull request:
changelog: Add `must_use_candidate` lint,  add `must-use-unit` lint, add `double_must_use` lint

The first one checks if an public function or method has no mutable argument and mutates no non-local data and lints if it has no `#[must_use]` attribute. It will skip inner functions, because those are usually highly local and the attribute doesn't have as much benefit there.

The second lints `#[must_use]` attributes on functions and methods that return unit. Those attributes are likely a remnant from a refactoring that removed the return value.

The last one lints for `#[must_use]` attributrs without text on functions that return a type which is already marked `#[must_use]`. This has no auto-suggestion, because while it would be easy to do, there may be value in writing a detailed text for the attribute instead.

This fixes #4526

phansch commented on tests/ui/must_use_pure.stderr:
```diff
@@ -0,0 +1,10 @@
+LL | pub fn pure(i: u8) -> u8 {
+   | ^^^^^^^^^^^^^^^^^^^^^^^^ help: add the attribute: `#[must_use] pub fn pure(i: u8) -> u8`
+   |
+   = note: `-D clippy::pure-without-must-use` implied by `-D warnings`
```
It would be nice to have more pure function examples that trigger the lint. Does it trigger in pure trait methods, for example?

llogiq commented:
The error seems unrelated? Perhaps a Rust PR that changed the error output.

phansch commented:
> The error seems unrelated? Perhaps a Rust PR that changed the error output.

Yup, that was fixed by https://github.com/rust-lang/rust-clippy/pull/4565

llogiq commented on clippy_lints/src/functions.rs:
```diff
@@ -84,19 +85,75 @@ declare_clippy_lint! {
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for pure functions that have no `#[must_use]`
+    /// attribute.
```
Good idea

llogiq commented:
I adressed your comments, @phansch. Thank you for the thorough review.

llogiq commented:
Added the tests and fixed the doc nit.

sinkuu commented on clippy_lints/src/functions.rs:
```diff
@@ -268,6 +390,138 @@ impl<'a, 'tcx> Functions {
+                return false;
+            }
+            !KNOWN_IMMUTABLES.iter().any(|path| match_def_path(cx, adt.did, path))
+                && KNOWN_MUTABLES.iter().any(|path| match_def_path(cx, adt.did, path))
```
Can [`TyS::is_freeze`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TyS.html#method.is_freeze) be used here?

llogiq replied:
Great idea! That catches many more instances than my hacky logic. Need to fix the fallout, though. :smiley:

phansch commented on CHANGELOG.md:
```diff
@@ -1065,6 +1065,7 @@ Released 2018-09-13
 [`modulo_one`]: https://rust-lang.github.io/rust-clippy/master/index.html#modulo_one
 [`multiple_crate_versions`]: https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions
 [`multiple_inherent_impl`]: https://rust-lang.github.io/rust-clippy/master/index.html#multiple_inherent_impl
+[`must_use_unit`]: https://rust-lang.github.io/rust-clippy/master/index.html#must_use_unit
```
Should this maybe be called `needless_must_use`? It would be more in line with our other `needless_*` lints.

llogiq replied:
I'm not married to the name, but I'd expect a `needless_must_use` lint to check for more than the a unit result. On the other hand, I cannot think of anything else to check, so maybe that's a moot point.

llogiq replied:
I now have two lints, one is called `must_use_unit` and the other is called `double_must_use`.

Centril commented on clippy_lints/src/functions.rs:
```diff
@@ -144,35 +205,96 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Functions {
+                if attr_span.is_none()
+                    && self.depth == 1
+                    && !returns_unit(&sig.decl)
+                    && !in_external_macro(cx.sess(), item.span)
```
This check with the two `&&`s are duplicated 3 times.

llogiq replied:
I changed the code & put it in a function.

Centril commented on clippy_lints/src/functions.rs:
```diff
@@ -144,35 +205,96 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Functions {
+                        cx,
+                        body,
+                        item.span,
+                        "this looks like a pure method without `#[must_use]` attribute",
```
The message here is repeated and could be in a `const`.

llogiq replied:
I changed the code so the message is no longer duplicated.

Centril commented:
I think this PR advocates a rather strange definition of "pure" which is not referentially transparent nor even deterministic. Looking at the diff it seems to have caused a non-trivial amount of false positives, e.g. involving file opening and such. Focusing on mutable types is also not enough, you need to consider side-effects though functions or basic things like `println!()`. Also, this seems like an analysis that is best done on MIR rather than HIR and should be more like the `const fn` analyses of the compiler.

llogiq commented:
Purity is – perhaps even in FP circles – a nebulous concept. Some people argue that it's about freedom from side effects (yet there is no such thing – any program running on a computer will burn cycles as a side effect, for example), others say they need to be referentially transparent or even deterministic. This lint takes a pragmatic approach. If a function does not mutate a static or its arguments, you'll likely want to call it for its result. I personally don't think file opening is a side effect I'd call a function for if I didn't want to do something with the file.

Also I agree that the lint triggers on a lot of function. Can you point to some you consider false positives?

mikerite commented:
This lint shouldn't apply to functions returning `Result` as you get a `rustc` warning if you don't use the result anyway.

mikerite commented:
`Iter` as well. Maybe there is a way to look up if a struct is `must_use`.

Centril commented on clippy_dev/src/lib.rs:
```diff
@@ -137,6 +137,7 @@ pub fn gen_deprecated(lints: &[Lint]) -> Vec<String> {
 /// Gathers all files in `src/clippy_lints` and gathers all lints inside
+#[must_use]
 pub fn gather_all() -> impl Iterator<Item = Lint> {
     lint_files().flat_map(|f| gather_from_file(&f))
```
Not a pure function, results in opening a file and reading from it.

Centril commented:
> Purity is – perhaps even in FP circles – a nebulous concept.

See https://pdfs.semanticscholar.org/32ff/28cb5b5c990fe7bdca510040ea22f5c13316.pdf for a well defined definition.

> Some people argue that it's about freedom from side effects (yet there is no such thing – any program running on a computer will burn cycles as a side effect, for example),

When reasoning about computer programs we do so preferably considering an idealized abstract machine in which some things (e.g. heat from burning cycles or bit flips due to radiation) are not observable. Otherwise it would be impossible to do all sorts of other reasonings as well (e.g. about soundness).

> [...] others say they need to be referentially transparent or even deterministic.

Referentially transparent is a common notion, but a basic property a pure function needs to satisfy is that given `f : A → B` then `∀ x, y: A. x ≡ y → f(x) ≡ f(y)`. This is not an operational definition, but is a sufficient yardstick to decide that the functions I've highlighted above are not pure.

> This lint takes a pragmatic approach. If a function does not mutate a static or its arguments, you'll likely want to call it for its result.

That's fine; I would suggest renaming the lint to something more suggestive and to extend the "known problems" to note more false positives. In my view, a better, but still pragmatic solution would be to do this on MIR.

> I personally don't think file opening is a side effect I'd call a function for if I didn't want to do something with the file.

Merely opening a file is not an interesting side-effect (yet it might cause a problem if the file was already open), but reading it means you are reading the state of the world and that can result in indeterministic results.

llogiq commented:
@Centril I see what you mean now. I'll best implement @mikerite's suggestion and rename the lint to `must_use_candidate`. It's already a pedantic lint, so false positives are somewhat acceptable.

llogiq commented:
Is there a simple way to get the return `ty::Ty` of a function/method? Trying to get the `cx.tables.liberated_fn_sigs()` and indexing on the `hir_id` of the function/method doesn't work.

bors commented:
:umbrella: The latest upstream changes (presumably #4574) made this pull request unmergeable. Please resolve the merge conflicts.

sinkuu commented on clippy_lints/src/functions.rs:
```diff
@@ -268,6 +390,167 @@ impl<'a, 'tcx> Functions {
+    }
+}
+
+static KNOWN_WRAPPER_TYS: &[&[&str]] = &[&["alloc", "rc", "Rc"], &["std", "sync", "Arc"]];
```
Both `Arc` and `Rc` contains `NonNull` ("`*mut T` but non-zero" type) as a field. Can you check for it instead?

sinkuu commented on clippy_lints/src/functions.rs:
```diff
@@ -331,3 +615,72 @@ impl<'a, 'tcx> DerefVisitor<'a, 'tcx> {
+            Assign(ref assignee, _) | AssignOp(_, ref assignee, _) => {
+                self.mutates_static |= is_mutated_static(self.cx, assignee)
+            },
+            _ => {},
```
How about checking for `&mut STATIC`?

llogiq replied:
Added.

phansch commented:
> Is there a simple way to get the return `ty::Ty` of a function/method? Trying to get the `cx.tables.liberated_fn_sigs()` and indexing on the `hir_id` of the function/method doesn't work.

`return_ty` in `utils/mod.rs` should get you the fn return type, you'll just have to pass the correct hir_id.

bors commented:
:umbrella: The latest upstream changes (presumably #4592) made this pull request unmergeable. Please resolve the merge conflicts.

llogiq commented:
This should now be ready to merge, I think (unless another new lint gets merged first, in which case I'll need to rebase again).

bors commented:
:umbrella: The latest upstream changes (presumably #4615) made this pull request unmergeable. Please resolve the merge conflicts.

llogiq commented:
@flip1995 r?

phansch commented:
I think `check_needless_must_use` is missing an external macro check?

Could you try adding [this diff](https://gist.github.com/phansch/b7e42d3bba55def4268abd972201b2f2) and see what happens? (Unfortunately my locally Clippy is currently not working, so I'm not sure if that diff compiles already)

llogiq commented:
@phansch first, isn't that missing a `#[must_use]` annotation for `foo()`? Also why would we want to quell the lint in macros? I can see how this would be a good idea for the `double_must_use` lint though.

llogiq commented:
@phansch I've added your test case as-is and also added `in_external_macro` checks to `must_use_unit` and `double_must_use`. Unfortunately, CI is broken again, we need yet another rustup.

phansch commented:
> isn't that missing a #[must_use] annotation for foo()

oh, yes! :sweat_smile: 

> Also why would we want to quell the lint in macros?

The test is just there to check that we don't lint occurences in external macros. It should still lint for non-external macros.

phansch commented on clippy_lints/src/functions.rs:
```diff
@@ -268,6 +392,160 @@ impl<'a, 'tcx> Functions {
+    });
+}
+
+fn must_use_attr(attrs: &[Attribute]) -> Option<&Attribute> {
```
I believe you can replace this function with:

```rust
syntax::attr::contains_name(attrs, sym!(must_use)
```

llogiq replied:
That returns only `bool`, I need the attribute for its node and span later.

phansch commented on clippy_lints/src/functions.rs:
```diff
@@ -144,35 +226,77 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Functions {
 
 impl<'a, 'tcx> Functions {
     fn check_arg_number(self, cx: &LateContext<'_, '_>, decl: &hir::FnDecl, span: Span) {
-        // Remove the function body from the span. We can't use `SourceMap::def_span` because the
```
It would be nice to keep this comment in the extracted function, too

llogiq replied:
I think I have a better option; I will need some time to apply it though.

llogiq replied:
I removed the function, there's a better solution anyway.

llogiq commented:
Let's see what CI thinks about it. I may need yet another rebase though.

llogiq commented:
@phansch r?

phansch commented:
@bors r+ nice!

bors commented:
:pushpin: Commit cc622608db7318b1c0fe3ccd541558436c7c6c4c has been approved by `phansch`

<!-- @bors r=phansch cc622608db7318b1c0fe3ccd541558436c7c6c4c -->
<!-- homu: {"type":"Approved","sha":"cc622608db7318b1c0fe3ccd541558436c7c6c4c","approver":"phansch"} -->

bors commented:
:hourglass: Testing commit cc622608db7318b1c0fe3ccd541558436c7c6c4c with merge 44ee6641ab183920ff32b9286501d6df922e87c8...
<!-- homu: {"type":"BuildStarted","head_sha":"cc622608db7318b1c0fe3ccd541558436c7c6c4c","merge_sha":"44ee6641ab183920ff32b9286501d6df922e87c8"} -->

phansch commented:
@bors retry (yielding priority to #4663)

bors commented:
:hourglass: Testing commit cc622608db7318b1c0fe3ccd541558436c7c6c4c with merge c72c428981142e3aad72095e3b06b7bc10bb057a...
<!-- homu: {"type":"BuildStarted","head_sha":"cc622608db7318b1c0fe3ccd541558436c7c6c4c","merge_sha":"c72c428981142e3aad72095e3b06b7bc10bb057a"} -->

phansch commented:
@bors retry

bors commented:
:hourglass: Testing commit cc622608db7318b1c0fe3ccd541558436c7c6c4c with merge 8fae2dd3c1bfed13bdd6c0cfd4170dd1363f25f7...
<!-- homu: {"type":"BuildStarted","head_sha":"cc622608db7318b1c0fe3ccd541558436c7c6c4c","merge_sha":"8fae2dd3c1bfed13bdd6c0cfd4170dd1363f25f7"} -->

bors commented:
:sunny: Test successful - [checks-travis](https://travis-ci.com/rust-lang/rust-clippy/builds/131841355), [status-appveyor](https://ci.appveyor.com/project/rust-lang-libs/rust-clippy/builds/28100361)
Approved by: phansch
Pushing 8fae2dd3c1bfed13bdd6c0cfd4170dd1363f25f7 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"phansch","base_ref":"master","builders":{"status-appveyor":"https://ci.appveyor.com/project/rust-lang-libs/rust-clippy/builds/28100361","checks-travis":"https://travis-ci.com/rust-lang/rust-clippy/builds/131841355"},"merge_sha":"8fae2dd3c1bfed13bdd6c0cfd4170dd1363f25f7"} -->

llogiq commented:
Thanks, @phansch!
